{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Evidence Alignment Analysis\n",
    "\n",
    "### Data Ingestion\n",
    "- load the evidence alignment results\n",
    "- unpack the experiment config results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from typing import List, Dict\n",
    "import json\n",
    "import pandas as pd\n",
    "\n",
    "\n",
    "LOG_DIR = \"/home/ubuntu/DSWizard/experiment_logs\"\n",
    "HYPOTHESIS_DIR = \"/home/ubuntu/DSWizard/benchmark_datasets/cBioPortal/hypothesis\"\n",
    "OUTPUT_DIR = \"/home/ubuntu/DSWizard/eval_evidence_alignment\"\n",
    "COMPLETED_TASKS_FILE = os.path.join(OUTPUT_DIR, \"completed_tasks.txt\")\n",
    "\n",
    "EVIDENCE_ALIGNMENT_RESULTS_FILE = os.path.join(OUTPUT_DIR, \"eval_results.json\")\n",
    "\n",
    "df = pd.read_json(EVIDENCE_ALIGNMENT_RESULTS_FILE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# unpack the experiment_config column into separate columns\n",
    "\n",
    "df = pd.concat([df.drop('experiment_config', axis=1), df['experiment_config'].apply(pd.Series)], axis=1)\n",
    "\n",
    "# extract the agent type, which is composed of the agent_type, and it's hyperparameters\n",
    "def get_agent_type(row: Dict) -> str:\n",
    "    \"\"\"\n",
    "    Format agent type based on configuration.\n",
    "    For react agent: (react, step_count)\n",
    "    For reasoning coder: (reasoning_coder, planning_model, coding_model)\n",
    "    For coder: (coder, model_name)\n",
    "    \"\"\"\n",
    "    agent_type = row[\"agent_type\"]\n",
    "    if agent_type == \"react\":\n",
    "        return f\"(react, {row['step_count']}, {row['model_name']})\"\n",
    "    elif agent_type == \"reasoning_coder\":\n",
    "        return f\"(reasoning_coder, {row['planning_model']}, {row['coding_model']})\"\n",
    "    elif agent_type == \"coder\":\n",
    "        return f\"(coder, {row['model_name']})\"\n",
    "    elif agent_type == \"reasoning_react\":\n",
    "        return f\"(reasoning_react, {row['plan_model_name']}, {row['agent_model_name']}, {row['step_count']})\"\n",
    "    return agent_type\n",
    "\n",
    "\n",
    "df['agent_summary'] = df.apply(get_agent_type, axis=1)\n",
    "# display(df['agent_summary'].value_counts())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Summarize Alignment Results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING: Length mismatch between ground_truth_evidence and alignment_results for row ground_truth_evidence      [GNH DPMs showed a higher frequency of TP53 al...\n",
      "generated_evidence         [Evidence 1: The frequency of TP53 alterations...\n",
      "eval_evidence_alignment    [{'evidence_id': '0', 'alignment': 'Supported'...\n",
      "agent_type                                                   reasoning_coder\n",
      "api_type                                                               azure\n",
      "language                                                              python\n",
      "final_response_model                                                  gpt-4o\n",
      "model_name                                                            gpt-4o\n",
      "step_count                                                               NaN\n",
      "pmid                                                                38630790\n",
      "hypothesis_index                                                           1\n",
      "dataset_ids                                                [plmeso_msk_2024]\n",
      "planning_model                                                       o3-mini\n",
      "coding_model                                                          gpt-4o\n",
      "agent_summary                             (reasoning_coder, o3-mini, gpt-4o)\n",
      "Name: 6217, dtype: object\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ground_truth_evidence</th>\n",
       "      <th>generated_evidence</th>\n",
       "      <th>eval_evidence_alignment</th>\n",
       "      <th>agent_type</th>\n",
       "      <th>api_type</th>\n",
       "      <th>language</th>\n",
       "      <th>final_response_model</th>\n",
       "      <th>model_name</th>\n",
       "      <th>step_count</th>\n",
       "      <th>pmid</th>\n",
       "      <th>hypothesis_index</th>\n",
       "      <th>dataset_ids</th>\n",
       "      <th>planning_model</th>\n",
       "      <th>coding_model</th>\n",
       "      <th>agent_summary</th>\n",
       "      <th>alignment_eval_supported</th>\n",
       "      <th>alignment_eval_contradicted</th>\n",
       "      <th>alignment_eval_missed</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>[The disparity in genomic stability is more pr...</td>\n",
       "      <td>[Initial data on structural variation grouped ...</td>\n",
       "      <td>[{'evidence_id': '0', 'alignment': 'Missed'}]</td>\n",
       "      <td>react</td>\n",
       "      <td>azure</td>\n",
       "      <td>python</td>\n",
       "      <td>gpt-4o</td>\n",
       "      <td>gpt-4o</td>\n",
       "      <td>16.0</td>\n",
       "      <td>32015526</td>\n",
       "      <td>1</td>\n",
       "      <td>[luad_oncosg_2020]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>(react, 16.0, gpt-4o)</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>[10.5% of myxofibrosarcomas have NF1 mutations...</td>\n",
       "      <td>[No samples of Myxofibrosarcoma and Pleomorphi...</td>\n",
       "      <td>[{'evidence_id': '0', 'alignment': 'Contradict...</td>\n",
       "      <td>coder</td>\n",
       "      <td>azure</td>\n",
       "      <td>python</td>\n",
       "      <td>gpt-4o</td>\n",
       "      <td>gpt-4o</td>\n",
       "      <td>NaN</td>\n",
       "      <td>20601955</td>\n",
       "      <td>1</td>\n",
       "      <td>[sarc_mskcc]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>(coder, gpt-4o)</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>[Higher frequency of KIT mutations in Triple-W...</td>\n",
       "      <td>[The data provided indicates that within the T...</td>\n",
       "      <td>[{'evidence_id': '0', 'alignment': 'Contradict...</td>\n",
       "      <td>coder</td>\n",
       "      <td>azure</td>\n",
       "      <td>python</td>\n",
       "      <td>gpt-4o</td>\n",
       "      <td>gpt-4o</td>\n",
       "      <td>NaN</td>\n",
       "      <td>26091043</td>\n",
       "      <td>1</td>\n",
       "      <td>[skcm_tcga_pub_2015]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>(coder, gpt-4o)</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>[Microsatellite instability-high status was en...</td>\n",
       "      <td>[The contingency table only contains data for ...</td>\n",
       "      <td>[{'evidence_id': '0', 'alignment': 'Missed'}]</td>\n",
       "      <td>react</td>\n",
       "      <td>azure</td>\n",
       "      <td>python</td>\n",
       "      <td>gpt-4o</td>\n",
       "      <td>o3-mini</td>\n",
       "      <td>16.0</td>\n",
       "      <td>35235413</td>\n",
       "      <td>2</td>\n",
       "      <td>[crc_dd_2022]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>(react, 16.0, o3-mini)</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>[HiC subtype is associated with TP53 mutation.]</td>\n",
       "      <td>[Data was collected and merged from 'data_clin...</td>\n",
       "      <td>[{'evidence_id': '0', 'alignment': 'Contradict...</td>\n",
       "      <td>react</td>\n",
       "      <td>azure</td>\n",
       "      <td>python</td>\n",
       "      <td>gpt-4o</td>\n",
       "      <td>gpt-4o</td>\n",
       "      <td>16.0</td>\n",
       "      <td>25583476</td>\n",
       "      <td>1</td>\n",
       "      <td>[egc_tmucih_2015]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>(react, 16.0, gpt-4o)</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10234</th>\n",
       "      <td>[Microsatellite instability-high status was en...</td>\n",
       "      <td>[The dataset loaded does not contain explicit ...</td>\n",
       "      <td>[{'evidence_id': '0', 'alignment': 'Missed'}]</td>\n",
       "      <td>react</td>\n",
       "      <td>azure</td>\n",
       "      <td>python</td>\n",
       "      <td>gpt-4o</td>\n",
       "      <td>gpt-4o</td>\n",
       "      <td>16.0</td>\n",
       "      <td>35235413</td>\n",
       "      <td>2</td>\n",
       "      <td>[crc_dd_2022]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>(react, 16.0, gpt-4o)</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10235</th>\n",
       "      <td>[Presence of truncating mutations in RBM10 and...</td>\n",
       "      <td>[Count of Truncating Mutations per Gene shows ...</td>\n",
       "      <td>[{'evidence_id': '0', 'alignment': 'Supported'}]</td>\n",
       "      <td>reasoning_coder</td>\n",
       "      <td>azure</td>\n",
       "      <td>python</td>\n",
       "      <td>gpt-4o</td>\n",
       "      <td>gpt-4o</td>\n",
       "      <td>NaN</td>\n",
       "      <td>22980975</td>\n",
       "      <td>1</td>\n",
       "      <td>[luad_broad]</td>\n",
       "      <td>o3-mini</td>\n",
       "      <td>gpt-4o</td>\n",
       "      <td>(reasoning_coder, o3-mini, gpt-4o)</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10236</th>\n",
       "      <td>[Predictions from kinase-substrate network ana...</td>\n",
       "      <td>[The regression model assessing the prediction...</td>\n",
       "      <td>[{'evidence_id': '0', 'alignment': 'Contradict...</td>\n",
       "      <td>coder</td>\n",
       "      <td>azure</td>\n",
       "      <td>python</td>\n",
       "      <td>gpt-4o</td>\n",
       "      <td>o3-mini</td>\n",
       "      <td>NaN</td>\n",
       "      <td>32888432</td>\n",
       "      <td>4</td>\n",
       "      <td>[coadread_cass_2020]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>(coder, o3-mini)</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10237</th>\n",
       "      <td>[TP53 alterations occur in half of dysplasia c...</td>\n",
       "      <td>[A high proportion (79.64%) of colitis samples...</td>\n",
       "      <td>[{'evidence_id': '0', 'alignment': 'Missed'}]</td>\n",
       "      <td>reasoning_coder</td>\n",
       "      <td>azure</td>\n",
       "      <td>python</td>\n",
       "      <td>gpt-4o</td>\n",
       "      <td>gpt-4o</td>\n",
       "      <td>NaN</td>\n",
       "      <td>36611031</td>\n",
       "      <td>0</td>\n",
       "      <td>[bowel_colitis_msk_2022]</td>\n",
       "      <td>o3-mini</td>\n",
       "      <td>gpt-4o</td>\n",
       "      <td>(reasoning_coder, o3-mini, gpt-4o)</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10238</th>\n",
       "      <td>[Patterns of MSI and mutations in 7 genes show...</td>\n",
       "      <td>[The mutation matrix and dendrogram were gener...</td>\n",
       "      <td>[{'evidence_id': '0', 'alignment': 'Contradict...</td>\n",
       "      <td>coder</td>\n",
       "      <td>azure</td>\n",
       "      <td>python</td>\n",
       "      <td>gpt-4o</td>\n",
       "      <td>o3-mini</td>\n",
       "      <td>NaN</td>\n",
       "      <td>28485815</td>\n",
       "      <td>1</td>\n",
       "      <td>[uccc_nih_2017]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>(coder, o3-mini)</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10239 rows Ã— 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   ground_truth_evidence  \\\n",
       "0      [The disparity in genomic stability is more pr...   \n",
       "1      [10.5% of myxofibrosarcomas have NF1 mutations...   \n",
       "2      [Higher frequency of KIT mutations in Triple-W...   \n",
       "3      [Microsatellite instability-high status was en...   \n",
       "4        [HiC subtype is associated with TP53 mutation.]   \n",
       "...                                                  ...   \n",
       "10234  [Microsatellite instability-high status was en...   \n",
       "10235  [Presence of truncating mutations in RBM10 and...   \n",
       "10236  [Predictions from kinase-substrate network ana...   \n",
       "10237  [TP53 alterations occur in half of dysplasia c...   \n",
       "10238  [Patterns of MSI and mutations in 7 genes show...   \n",
       "\n",
       "                                      generated_evidence  \\\n",
       "0      [Initial data on structural variation grouped ...   \n",
       "1      [No samples of Myxofibrosarcoma and Pleomorphi...   \n",
       "2      [The data provided indicates that within the T...   \n",
       "3      [The contingency table only contains data for ...   \n",
       "4      [Data was collected and merged from 'data_clin...   \n",
       "...                                                  ...   \n",
       "10234  [The dataset loaded does not contain explicit ...   \n",
       "10235  [Count of Truncating Mutations per Gene shows ...   \n",
       "10236  [The regression model assessing the prediction...   \n",
       "10237  [A high proportion (79.64%) of colitis samples...   \n",
       "10238  [The mutation matrix and dendrogram were gener...   \n",
       "\n",
       "                                 eval_evidence_alignment       agent_type  \\\n",
       "0          [{'evidence_id': '0', 'alignment': 'Missed'}]            react   \n",
       "1      [{'evidence_id': '0', 'alignment': 'Contradict...            coder   \n",
       "2      [{'evidence_id': '0', 'alignment': 'Contradict...            coder   \n",
       "3          [{'evidence_id': '0', 'alignment': 'Missed'}]            react   \n",
       "4      [{'evidence_id': '0', 'alignment': 'Contradict...            react   \n",
       "...                                                  ...              ...   \n",
       "10234      [{'evidence_id': '0', 'alignment': 'Missed'}]            react   \n",
       "10235   [{'evidence_id': '0', 'alignment': 'Supported'}]  reasoning_coder   \n",
       "10236  [{'evidence_id': '0', 'alignment': 'Contradict...            coder   \n",
       "10237      [{'evidence_id': '0', 'alignment': 'Missed'}]  reasoning_coder   \n",
       "10238  [{'evidence_id': '0', 'alignment': 'Contradict...            coder   \n",
       "\n",
       "      api_type language final_response_model model_name  step_count      pmid  \\\n",
       "0        azure   python               gpt-4o     gpt-4o        16.0  32015526   \n",
       "1        azure   python               gpt-4o     gpt-4o         NaN  20601955   \n",
       "2        azure   python               gpt-4o     gpt-4o         NaN  26091043   \n",
       "3        azure   python               gpt-4o    o3-mini        16.0  35235413   \n",
       "4        azure   python               gpt-4o     gpt-4o        16.0  25583476   \n",
       "...        ...      ...                  ...        ...         ...       ...   \n",
       "10234    azure   python               gpt-4o     gpt-4o        16.0  35235413   \n",
       "10235    azure   python               gpt-4o     gpt-4o         NaN  22980975   \n",
       "10236    azure   python               gpt-4o    o3-mini         NaN  32888432   \n",
       "10237    azure   python               gpt-4o     gpt-4o         NaN  36611031   \n",
       "10238    azure   python               gpt-4o    o3-mini         NaN  28485815   \n",
       "\n",
       "       hypothesis_index               dataset_ids planning_model coding_model  \\\n",
       "0                     1        [luad_oncosg_2020]            NaN          NaN   \n",
       "1                     1              [sarc_mskcc]            NaN          NaN   \n",
       "2                     1      [skcm_tcga_pub_2015]            NaN          NaN   \n",
       "3                     2             [crc_dd_2022]            NaN          NaN   \n",
       "4                     1         [egc_tmucih_2015]            NaN          NaN   \n",
       "...                 ...                       ...            ...          ...   \n",
       "10234                 2             [crc_dd_2022]            NaN          NaN   \n",
       "10235                 1              [luad_broad]        o3-mini       gpt-4o   \n",
       "10236                 4      [coadread_cass_2020]            NaN          NaN   \n",
       "10237                 0  [bowel_colitis_msk_2022]        o3-mini       gpt-4o   \n",
       "10238                 1           [uccc_nih_2017]            NaN          NaN   \n",
       "\n",
       "                            agent_summary  alignment_eval_supported  \\\n",
       "0                   (react, 16.0, gpt-4o)                         0   \n",
       "1                         (coder, gpt-4o)                         0   \n",
       "2                         (coder, gpt-4o)                         0   \n",
       "3                  (react, 16.0, o3-mini)                         0   \n",
       "4                   (react, 16.0, gpt-4o)                         0   \n",
       "...                                   ...                       ...   \n",
       "10234               (react, 16.0, gpt-4o)                         0   \n",
       "10235  (reasoning_coder, o3-mini, gpt-4o)                         1   \n",
       "10236                    (coder, o3-mini)                         0   \n",
       "10237  (reasoning_coder, o3-mini, gpt-4o)                         0   \n",
       "10238                    (coder, o3-mini)                         0   \n",
       "\n",
       "       alignment_eval_contradicted  alignment_eval_missed  \n",
       "0                                0                      1  \n",
       "1                                3                      0  \n",
       "2                                2                      0  \n",
       "3                                0                      1  \n",
       "4                                1                      0  \n",
       "...                            ...                    ...  \n",
       "10234                            0                      1  \n",
       "10235                            0                      0  \n",
       "10236                            1                      0  \n",
       "10237                            0                      1  \n",
       "10238                            1                      0  \n",
       "\n",
       "[10239 rows x 18 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# create a summary of the alignment results\n",
    "def summarize_alignment_results(row: Dict) -> str:\n",
    "    \"\"\"\n",
    "    Summarize the alignment results for a given row.\n",
    "    \"\"\"\n",
    "    ground_truth_evidence = row['ground_truth_evidence']\n",
    "    generated_evidence = row['generated_evidence']\n",
    "    alignment_results = row['eval_evidence_alignment']\n",
    "    \n",
    "    if not len(ground_truth_evidence) == len(alignment_results):\n",
    "        print(f\"WARNING: Length mismatch between ground_truth_evidence and alignment_results for row {row}\")\n",
    "        # print(\"ground_truth_evidence: \", ground_truth_evidence)\n",
    "        # print(\"alignment_results: \", alignment_results)\n",
    "        # print(\"generated_evidence: \", generated_evidence)\n",
    "        # raise ValueError(\"Length mismatch between ground_truth_evidence and alignment_results\")\n",
    "    \n",
    "    values = {\n",
    "        \"supported\": 0,\n",
    "        \"contradicted\": 0,\n",
    "        \"missed\": 0\n",
    "    }\n",
    "    for i in range(len(ground_truth_evidence)):\n",
    "        res = alignment_results[i]['alignment']\n",
    "        res = res.strip().lower()\n",
    "        \n",
    "        if res not in ['supported', 'contradicted', 'missed']:\n",
    "            raise ValueError(f\"Invalid alignment result: {res}\")\n",
    "        \n",
    "        values[res] += 1\n",
    "        \n",
    "    for key, value in values.items():\n",
    "        row[f\"alignment_eval_{key}\"] = value\n",
    "        \n",
    "    return row\n",
    "\n",
    "df = df.apply(summarize_alignment_results, axis=1)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "DSWizard-u3vtC4yL",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
